enhancements/olm-deletes-operands.md

```diff
@@ -0,0 +1,209 @@
+
+The CSV is the more appropriate point to configure the cleanup option for an operator as the CSV represents the actual installation of the operator, whereas the Subscription only represents the intent to subscribe to updates from the catalog. Moreover, since deleting the CSV causes the operator to be removed, deleting the CSV should be the trigger to initiate operand cleanup. See the [section on CSV finalizers][finalizers-on-csvs] for more details.
+
+#### Deleting CRs
+
+The operand cleanup process, simply put, is the removal of all CRs being reconciled by an operator so it can run finalizers for each CR and clean up any associated operator-created resources.
+
+From OLM's perspective these would be all CRs in the `OperatorGroup`'s [target namespaces][target-namespaces] that are of the same type as the [owned APIservices][owned-apiservices] listed in the CSV. This represents the set of all CRs that the operator reconciles and is responsible for cleaning up after.
+
+OLM should not delete CRs of the type [required APIService][required-apiservices] as those are likely created by the operator, and whose cleanup should be handled by the operator's finalizer.
+
+#### Finalizers on CSVs
+
+OLM can initiate the process of operand clean up when the operator is uninstalled, which is when the CSV is deleted. However, deleting a CSV causes the operator deployment (owned by the CSV) to be garbage collected.
+
+So in order to keep the operator around to process CR deletion events, OLM can add a finalizer to the CSV to prevent the CSV and the operator from being removed. Once OLM has completed operand cleanup and can verify the removal of all CRs, the CSV finalizer can be cleared so that the operator can be safely removed.
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+...
+metadata:
+  ...
+  finalizers:
+  - olm/operand-cleanup
+```
+
+In the event of operand cleanup being stalled due to some operator error, OLM will not clear the CSV finalizer so that the operator stays alive indefinitely. This way the operator can be observed to figure out the required action to unblock cleanup.
+
+#### Operand cleanup API
+
+The CSV spec can have an optional field `spec.cleanup` which lets a user configure the cleanup behaviour before the CSV is deleted.
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+...
+spec:
+  ...
+  cleanup:
+    enabled: false
+```
+
+With `spec.cleanup.enabled: true` OLM will first add the `olm/operand-cleanup` finalizer on the CSV. Now when the CSV is deleted by the user, OLM will delete all CRs for the operator's owned APIs, in the target namespaces of that operator. Once the CRs are successfully removed (after the operator clears its finalizers), OLM will clear the finalizer from the CSV to let the CSV and operator be uninstalled.
```

Comment:
I'm also wondering if we're missing something on the CSV spec here. Right now we are munging "does this operator even have this feature" with "we're going to actually delete stuff when you delete the CSV". Should we have a separate flag or annotation that explicitly states "my operator supports cleaning up operands" without also being the level trigger for turning that function on or off?

Reply:
I can see the value in having an annotation to signal if an operator has the capability to cleanup via finalizers.
Presumably the CSV author can add this annotation to CSV since they're likely the operator author as well.
And OLM users can't enable cleanup for operators/CSVs that haven't set this annotation (except by adding the annotation themselves I guess).

It does become a two step process now (1. set annotation 2. set `cleanup:true`) but if the two steps are meant to get opt-ins from different personas (operator author, olm user) then it's not redundant.

Reply:
Cleaning up Operands is not a feature only some Operators have, IMHO. It's a basic aspect of reconciliation. We would add an annotation for something that 99% of all known Operators packaged for OLM already do, because they were written with some derivate of controller-runtime.

Reply:
But they explicitly won't have that feature by default, right? If we add finalizers to the crs, the operators will need to have explicit code in their deletion reconcilers to remove those finalizers. Otherwise, those operators will be DOA. We could certainly make the sdk scaffold some of that code out, but that's not a guarantee for any arbitrary OLM operator.

Reply:
> If we add finalizers to the crs, the operators will need to have explicit code in their deletion reconcilers to remove those finalizers. Otherwise, those operators will be DOA. 

Yes, this is something I've been thinking about -- there is an implicit assumption for this enhancement that finalizers need to be implemented (correctly) on the operands and reconciled by the operator. If no finalizers are present then the CRs are immediately deleted. I think this needs to be documented and really well presented to users so we avoid any potential inadvertent data loss scenarios (big bold letters).

Reply:
From what I understand (and please correct me if I'm wrong) there are two cleanup scenarios:

The operator opts-in to cleanup - OLM adds its own `olm/operand-cleanup` finalizer to the CSV on creation - the CSV creates the operator deployment. The operator creates CRs that have their own `foo/bar` finalizer on them. When the CSV is deleted, OLM sets the deletion timestamp on the CRs, which causes the `foo/bar` finalizer to run. When the operator has finished processing and removing the finalizer from all the CRs, they are deleted. The `olm/operand-cleanup` finalizer is then removed, which causes the CSV to be deleted (and the operator gets gc'd). This is the normal happy path.

The second scenario is similar, but this time the operator creates the CRs with no finalizer on them. When OLM sets the deletion timestamp (by calling `Delete()`) the CRs get immediately deleted. The `olm/operand-cleanup` finalizer is then removed, which causes the CSV to be deleted (and the operator gets gc'd). 

The second scenario seems a bit more confusing -- without any sort of cleanup, wouldn't the CRs get gc'd if they are owned by the operator (which gets gc'd when the CSV is deleted)? It feels like it doesn't make sense to enable cleanup if no finalizer is present on the CRs, it just adds some complexity. 

Maybe the annotation can be a list of finalizers that the operator is using for its CRs to make it extremely clear what happens when it is deleted.

Reply:
> The operator creates CRs that have their own foo/bar finalizer on them

@exdx Operators don't create the primary CR or have ownership of them (or that's not a pattern that I've seen).
The operator user creates a CR for an application and the operator then sets finalizers on them.

The operator may create secondary CRs when reconciling for the primary CR, but even those secondary CRs, like other secondary resources would have their OwnerRef set to the primary CR.

> without any sort of cleanup, wouldn't the CRs get gc'd if they are owned by the operator

The operator is supposed to be stateless and shouldn't have ownership of any resource/CR that comprises the application.
So deleting the operator won't GC the CRs.

Although someone can correct me if there is an alternative CR/resource ownership model that I'm not aware of.

Reply:
> If we add finalizers to the crs, the operators will need to have explicit code in their deletion reconcilers to remove those finalizers

I also want to make this explicit since there seems to be some confusion. Under this proposal "we" or OLM is not going to add finalizers to the CRs of an application. That is the operator's responsibility.
OLM is only going to add finalizers to block deletion of the CSV until it can finish the cleanup of the application CRs.

I'll spell this out more clearly in case it wasn't clear from the wording of the proposal.

Reply:
> We would add an annotation for something that 99% of all known Operators packaged for OLM already do, because they were written with some derivate of controller-runtime.

> But they explicitly won't have that feature by default, right? 

 @dmesser I think @kevinrizza is correct here. Finalizers are a basic aspect of reconciliation but the controller-runtime/SDK/Kubebuilder does not force you or scaffold code for you to use finalizers. You have to do that yourself.
https://sdk.operatorframework.io/docs/building-operators/golang/advanced-topics/#handle-cleanup-on-deletion
https://book.kubebuilder.io/reference/using-finalizers.html

Which is why most operators probably don't have this implemented until they realize they need to do it.
Although I can't speak to how many operators actually use finalizers in practice I'd have to guess it's not most of them.

Reply:
Yeah, based on this discussion I think adding a spec field to the CSV to define that the operator has opted into the feature is needed. We can certainly reach out to @jmrodri and @joelanford about how they would imagine that could be scaffolded by the SDK when generating APIs as a separate step though.

Reply:
i'm a big plus one on "the operator must declare it supports cleanup" and "the admin does not get offered the option to cleanup operands on uninstall unless the operator declares it supprosts cleanup".

I've heard there are concerns that this might constrain adoption of the feature(because it requires action on the operator author's part), but i don't understand how the feature can be considered "adopted" w/o explicit operator support for it.  If OLM just deletes CRs on uninstall and the operator either doesn't have logic to delete the operand, or doesn't get a chance to run the logic (because OLM deletes the operator immediately after deleting the CRs), then that is worse than not doing anything (and worse than not offering the admin the option of having the CRs removed during uninstall).